amaranth/lib/data.py

```diff
@@ -789,6 +799,221 @@ def __repr__(self):
+    or from a bit-pattern using :meth:`Layout.from_bits`.
+
+    Accessing a const
+    #################
+
+    Slicing a const or accessing its attributes returns a part of the underlying value
+    corresponding
+    to the field with that index or name. If the shape of the field is a :class:`Layout`,
+    the accessed value will be another :class:`Const`. If it is another :ref:`shape-like <lang-shapelike>`
+    object implementing ``from_bits``, it will be the result of calling that method. Otherwise,
+    it will be a plain :class:`int`.
+
+    Slicing a view whose layout is an :class:`ArrayLayout` can be done with an index that is
+    an Amaranth value instead of a constant integer. The returned element is chosen dynamically
+    in that case, and the resulting value will be a :class:`View` instead of :class:`Const`.
+
+    A const can be compared for equality with another const or view of the same layout.
+    When compared with another const, the result will be a Python :class:`bool`.
+    When compared with a view, the result will be a single-bit value. No other operators
+    are supported on consts. If required, a const can be converted to an Amaranth value using
+    :meth:`as_value`, or to its underlying bit-pattern using :meth:`as_bits`.
+    """
+    def __init__(self, layout, target):
+        try:
+            cast_layout = Layout.cast(layout)
+        except TypeError as e:
+            raise TypeError(f"Const layout must be a layout, not {layout!r}") from e
+        try:
+            target = operator.index(target)
+        except TypeError as e:
+            raise TypeError(f"Const target must be an int, not {target!r}") from e
+        if target not in range(1 << cast_layout.size):
+            raise ValueError(f"Const target does not fit in {cast_layout.size} bit(s)")
+        for name, field in cast_layout:
+            if isinstance(name, str) and name[0] != "_" and hasattr(type(self), name):
+                warnings.warn("Const layout includes a field {!r} that will be shadowed by "
+                              "the const attribute '{}.{}.{}'"
+                              .format(name, type(self).__module__, type(self).__qualname__, name),
+                              SyntaxWarning, stacklevel=2)
+        self.__orig_layout = layout
+        self.__layout = cast_layout
+        self.__target = target
+
+    def shape(self):
+        """Get layout of this view.
+
+        Returns
+        -------
+        :class:`Layout`
+            The ``layout`` provided when constructing the view.
+        """
+        return self.__orig_layout
+
+    def as_value(self):
+        """Convert to Amaranth value.
+
+        Returns
+        -------
+        :class:`amaranth.hdl.Const`
+            A constant Amaranth value with the underlying bit-pattern.
+        """
+        return hdl.Const(self.__target, self.__layout.size)
+
+    def as_bits(self):
```

Comment:
Maybe add a warning to `View` that using a field called `as_bits` will break `Const`?

Reply:
It ended up a little awkward (I think the documentation will read better when we have custom `Const` subclass support?)